crates/wasi-common/src/snapshots/preview_1.rs

```diff
@@ -895,41 +934,67 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
-            if oflags.contains(OFlags::CREATE) {
-                required_caps = required_caps | DirCaps::CREATE_FILE;
-            }
+                || oflags.contains(OFlags::TRUNCATE))
+        {
+            return Err(Error::invalid_argument().context("directory oflags"));
+        }
 
-            let file_caps = dir_entry.child_file_caps(FileCaps::from(&fs_rights_base));
-            let dir = dir_entry.get_cap(required_caps)?;
-            let read = file_caps.contains(FileCaps::READ);
-            let write = file_caps.contains(FileCaps::WRITE)
-                || file_caps.contains(FileCaps::ALLOCATE)
-                || file_caps.contains(FileCaps::FILESTAT_SET_SIZE);
-            let file = dir
-                .open_file(symlink_follow, path.deref(), oflags, read, write, fdflags)
-                .await?;
-            drop(dir);
-            let fd = table.push(Box::new(FileEntry::new(file_caps, file)))?;
-            Ok(types::Fd::from(fd))
+        if oflags.contains(OFlags::DIRECTORY)
+            || fs_rights_base.contains(types::Rights::FD_READDIR)
+            || !(fs_rights_base.contains(types::Rights::FD_WRITE)
+                || oflags.contains(OFlags::CREATE)
+                || oflags.contains(OFlags::EXCLUSIVE)
+                || oflags.contains(OFlags::TRUNCATE))
+        {
+            match dir.open_dir(symlink_follow, path.deref()).await {
+                Ok(child_dir) => {
+                    let dir_caps = dir_entry.child_dir_caps(DirCaps::from(&fs_rights_base));
+                    let file_caps =
+                        dir_entry.child_file_caps(FileCaps::from(&fs_rights_inheriting));
+                    drop(dir);
+                    let fd = table.push(Box::new(DirEntry::new(
+                        dir_caps, file_caps, None, child_dir,
+                    )))?;
+                    return Ok(types::Fd::from(fd));
+                }
+                Err(e) => {
+                    // If we required a directory and didn't get one, fail.
+                    if oflags.contains(OFlags::DIRECTORY) {
+                        return Err(e);
+                    }
+                    // If the name didn't exist, or existed but isn't a
+                    // directory, proceed to try to open it as a file.
+                    // Otherwise fail.
+                    if !is_error_noent(&e) && !is_error_notdir(&e) {
+                        return Err(e);
+                    }
+                }
+            }
         }
+
+        let file_caps = dir_entry.child_file_caps(FileCaps::from(&fs_rights_base));
+        let dir = dir_entry.get_cap(required_caps)?;
+        let read = file_caps.contains(FileCaps::READ);
+        let write = file_caps.contains(FileCaps::WRITE)
+            || file_caps.contains(FileCaps::ALLOCATE)
+            || file_caps.contains(FileCaps::FILESTAT_SET_SIZE);
+        let file = dir
+            .open_file(symlink_follow, path.deref(), oflags, read, write, fdflags)
```

Comment:
You mentioned a race condition in https://github.com/bytecodealliance/wasmtime/pull/4947/files#r980506965 as reasoning for this PR, where the underlying filetype stored under a path could change between the calls, but doesn't this suffer from exact same race condition, where:
1. initial `open_dir` fails on path `A` with `ENOTDIR`
2. in the meantime, `A` is replaced by a directory
3. `open_file` fails, because `A` is now a directory
?

Reply:
That's a good point. I need to look at this more closely.